stdlib/dynarray.mli

```diff
@@ -0,0 +1,418 @@
+    of a dynamic array is proportional to the total number of elements
+    added or removed.
+    In other words, users need not care about capacity and reallocations,
+    and they will get reasonable behavior by default. However, in some
+    performance-sensitive scenarios the functions below can help control
+    memory usage or guarantee an optimal number of reallocations.
+*)
+
+val ensure_capacity : 'a t -> int -> unit
+(** [ensure_capacity a n] makes sure that [a] has capacity has least [n].
+
+    @raise Invalid_argument if the requested capacity is negative.
+    (We consider that this is a programming error.)
+
+    @raise Failure if the requested capacity is above
+      {!Sys.max_array_length}.
+    (We consider that this is a valid failure mode in some exceptional
+    scenarios. In particular, all functions adding elements to a dynamic
+    array may propagate this exception.)
+
+    An example use-case would be to implement [append_array]:
+{[
+    let append_array a arr =
+      ensure_capacity a (length a + Array.length arr);
+      Array.iter (fun v -> add_last a v) arr
+]}
+
+    Using [ensure_capacity] guarantees that at most one reallocation
+    will take place, instead of possibly several.
+
+    Without this [ensure_capacity] hint, the number of resizes would
+    be logarithmic in the length of [arr], creating a constant-factor
+    slowdown noticeable when [a] is small and [arr] is large.
+*)
+
+val fit_capacity : 'a t -> unit
+(** [fit_capacity a] shrinks the backing array so that its capacity is
+    exactly [length a], with no additional empty space at the
+    end. This can be useful to make sure there is no memory wasted on
+    a long-lived array.
+
+    Note that calling [fit_capacity] breaks the amortized complexity
+    guarantees provided by the default reallocation strategy, and may
+    result in more reallocations in the future.
+
+    If you know that a dynamic array has reached its final length,
+    which will remain fixed in the future, it is sufficient to call
+    [to_array] and only keep the resulting fixed-size
+    array. [fit_capacity] is useful when you need to keep a dynamic
+    array for eventual future resizes.
+*)
+
+val reset : 'a t -> unit
+(** [reset a] clears [a] and replaces its backing array by an empty array.
+
+    It is equivalent to [clear a; fit_capacity a].
+
+    Similar to {!Buffer.reset}. *)
+
+(** {2:noleaks No leaks: preservation of memory liveness}
+
+    The user-provided values reachable from a dynamic array [a] are
+    exactly the elements in the positions [0] to [length a - 1]. In
+    particular, no user-provided values are "leaked" by being present
+    in the backing array in position [length a] or later.
+*)
+
+(** {2:memory_layout Memory layout of dynarrays}
+
+    In the current implementation, the backing array of an
+    ['a Dynarray.t] is not an ['a array], but something closer to an
+    ['a option array] in terms of memory layout -- with a mutable field,
```

Comment:
I don't understand the sentence after the  `--`

Reply:
New attempt:

> In the current implementation, the backing array of an
> ['a Dynarray.t] is not an ['a array], but something with the same
> representation as an ['a option array] or ['a ref array].
> Each element is in a "box", allocated when the element is first
> added to the array -- see the implementation for more details.